File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    total = {} # 장르 : 총 재생 횟수
    album = {} # 장르 : [(플레이 횟수, 고유번호)]
    for i in range(len(genres)):
        total[genres[i]] = total.get(genres[i], 0) + plays[i]
        album[genres[i]] = album.get(genres[i], [])+[(plays[i], i)]

    sorted_total = sorted(total.items(), key=lambda x: x[1], reverse=True)

    for (genre, total) in sorted_total:
        album[genre] = sorted(album[genre], key=lambda x: (-x[0], x[1],))
        answer += [idx for (play, idx) in album[genre][:2]]

    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
import java.io.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        Map<String, Integer> cnt = new HashMap<>();
        Map<String, HashMap<Integer, Integer>> album = new HashMap<>();
        
        for(int i = 0; i < genres.length; i++){
            if(!cnt.containsKey(genres[i])){
                HashMap<Integer, Integer> m = new HashMap<>();
                m.put(i, plays[i]);
                album.put(genres[i], m);
                cnt.put(genres[i], plays[i]);
            } else {
                album.get(genres[i]).put(i, plays[i]);
                cnt.put(genres[i], cnt.get(genres[i]) + plays[i]);
                
            }
        }
        
        List<String> keySet = new ArrayList<>(cnt.keySet());
        keySet.sort((o1, o2) -> cnt.get(o2) - (cnt.get(o1)));
        List<Integer> answer = new ArrayList();
        
        for(String i : keySet) {
            HashMap<Integer, Integer> map = album.get(i);
            List<Integer> key = new ArrayList(map.keySet());
            
            key.sort((o1, o2) -> map.get(o2) - (map.get(o1)));
            
            answer.add(key.get(0));
            if(key.size() > 1)
                answer.add(key.get(1));
        }
        
        return answer.stream().mapToInt(i -> i).toArray();
    }
}